### PR TITLE
Storage/STG90-Data Lake Pagination Delete

### PR DESCRIPTION
**Pull request by microzchang**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_path_client.hpp [hunk 1]
```diff
@@ -306,8 +306,9 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
 
     /** @brief Http Pipeline */
     std::shared_ptr<Azure::Core::Http::_internal::HttpPipeline> m_pipeline;
-    /** @brief Customer provided encryption key. */
-    Azure::Nullable<EncryptionKey> m_customerProvidedKey;
+
+    /** @brief Client configurations*/
+    _detail::DatalakeClientConfiguration m_clientConfiguration;
```
[the remark this answers is not in the document]
Comment (microzchang) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_path_client.hpp:311]: Ok. I will add this to all clients in datalake.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_path_client.hpp [hunk 2]
```diff
@@ -306,25 +306,26 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
      */
     explicit DataLakePathClient(
         Azure::Core::Url pathUrl,
         Blobs::BlobClient blobClient,
         std::shared_ptr<Azure::Core::Http::_internal::HttpPipeline> pipeline,
-        Azure::Nullable<EncryptionKey> customerProvidedKey = Azure::Nullable<EncryptionKey>())
+        _detail::DatalakeClientConfiguration clientConfiguration)
         : m_pathUrl(std::move(pathUrl)), m_blobClient(std::move(blobClient)),
-          m_pipeline(std::move(pipeline)), m_customerProvidedKey(std::move(customerProvidedKey))
+          m_pipeline(std::move(pipeline)), m_clientConfiguration(clientConfiguration)
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_path_client.hpp:328]:
```suggestion
          m_pipeline(std::move(pipeline)), m_clientConfiguration(std::move(clientConfiguration))
```

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp [hunk 3]
```diff
@@ -145,7 +155,7 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
         builder,
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp:155]:
```suggestion
        std::move(builder),
```

File: sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp [hunk 4]
```diff
@@ -271,14 +283,29 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
-    _detail::PathClient::DeletePathOptions protocolLayerOptions;
-    protocolLayerOptions.LeaseId = options.AccessConditions.LeaseId;
-    protocolLayerOptions.IfMatch = options.AccessConditions.IfMatch;
-    protocolLayerOptions.IfNoneMatch = options.AccessConditions.IfNoneMatch;
-    protocolLayerOptions.IfModifiedSince = options.AccessConditions.IfModifiedSince;
-    protocolLayerOptions.IfUnmodifiedSince = options.AccessConditions.IfUnmodifiedSince;
-    protocolLayerOptions.Recursive = options.Recursive;
-    return _detail::PathClient::Delete(*m_pipeline, m_pathUrl, protocolLayerOptions, context);
+    bool paginated = m_clientConfiguration.ApiVersion >= "2023-08-03"
+        && m_clientConfiguration.TokenCredential != nullptr;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp:287]:
```suggestion
        && m_clientConfiguration.TokenCredential;
```

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp:287]: I think we can use the built-in `operator bool` converter.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp:287]: https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_bool

File: sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp [hunk 5]
```diff
@@ -271,14 +283,29 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+    std::string continuationToken = "";
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp:288]:
```suggestion
    std::string continuationToken;
```

File: sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp [hunk 6]
```diff
@@ -271,14 +283,29 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
+    do
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_path_client.cpp:289]: It's considered a bad practice to use `do while` loops in some style guides. Some formatter would emit a warning.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-do

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]

**Review by Jinming-Hu — APPROVED** [review]

**Review by Jinming-Hu — APPROVED** [review]

**Pull request merged**